### ISSUE TITLE
DependencyAnalysis misses dependency on scalar

### ISSUE THREAD

**Issue by sergisiso** [stfc/PSyclone#2727]
@hiker This is a simplified version of an issue we hit on NEMOv5, here my_val sometimes keeps its value for the next iteration, so it is essentially serialising the loop. Could we do something about this?
```
def test_new(fortran_reader):
    '''Tests can_loop_be_parallelised.'''
    source = '''program test
                integer :: i, my_val
                integer, dimension(10) :: array

                do i = 1, 10
                  if (array(i) > 3) then
                    my_val = array(i)
                    array(i) = my_val
                  else
                    array(i) = my_val
                  endif
                end do

                end program test'''

    psyir = fortran_reader.psyir_from_source(source)
    loop = psyir.children[0].children[0]
    dep_tools = DependencyTools()
    parallel = dep_tools.can_loop_be_parallelised(loop)
    assert parallel is False
```

**Comment by hiker** [comment]
I don't see an easy way :( I tried to rewrite it as something like:
```
do i=1, 10
    if (array(i)<=3) then
      inner: do j= i, 1, -1
          if (array(j)>3) then
              array(i) = array(j)
              break  inner
          endif
     enddo inner    ! not sure what to do if there is no element >3 :)
enddo
```
I believe that should roughly be the same logic - for each element <=3 find the previous element >3

But I assume the dependency analysis will still see read and write accesses to the same index in different iterations (i and j), and there actually is a race condition (consider 4, 1, 1,  in array... when the loop index 2 and 3 ... the 1s ... are executed in parallel, loop index 3 might read the value from loop 2, while loop 2 is writing ... ) except of course it doesn't matter, since the same value will be written in loop index 3 (the 4, either from array index 2 or array index 1).

But I think this rewrite should allow you to execute this in parallel (by disabling the DA). And of course, handling the problem what happens if the first element is <=3 ... my_val would be uninitialised, and the search loop won't find a value.

**Comment by sergisiso** [comment]
I would like to try to identify the dependency in the original code as this leads to compliable but code with different semantics (psyclone adds a private(my_var) since it finds the write-write), and it's very difficult to find out where the difference in results come from.

What make it hard? Is it the fact that it is a "scalar"? If I do something like:
```
    def test_new2(fortran_reader):
        '''Tests can_loop_be_parallelised.'''
        source = '''program test
                    integer :: i,j
                    integer, dimension(10) :: my_val
                    integer, dimension(10, 10) :: array

                    do j = 1, 10
                        do i = 1, 10
                          if (array(i, j) > 3) then
                            my_val(i) = 1
                            array(i, j) = my_val(i)
                          else
                            array(i, j) = my_val(i)
                          endif
                        end do
                    end do

                    end program test'''

        psyir = fortran_reader.psyir_from_source(source)
        loop = psyir.children[0].children[0]
        dep_tools = DependencyTools()

        parallel = dep_tools.can_loop_be_parallelised(loop)
        assert parallel is False
```
It properly detects the write-write dependency on my_val

**Comment by sergisiso** [comment]
In both cases, if in both condition bodies there was a `my_val( =` write (or `my_val(1) =`), they would be parallelisable with the `private(my_val)`.

Ignoring the scalar version for now, the problem seems to be that the write-write is not exhaustive to all branches, and therefore some are carried to the next iteration. Would it be possible to differentiate between:
- exhaustive write-write -> can be parallelised with a (first)private
- non-exhaustive write-write -> cannot be parallelised with a (first)private

**Comment by hiker** [comment]
@sergisiso What do you want the API to be? E.g. in a sequence like this:
```
   a=b   (1)
   if (i.eq.3) then   (2)
      c = d       (3)
   else
     e = f   (4)
   endif
```
We could first select the info the way we did before, meaning we get the access info for four statements (the three assignments and reading `i` in the if statement).

Then we could 'flatten' this - I.e. combine the info for (3) and (4) into 2, which can then indicate that there is a 'non-exhaustive' read of `d` and `f`, and similar writes to `c` and `e`. But how should I represent this? Should each variable access in there have an additional flag ('conditional` ... I like this more than exhaustive): so each read/write could be conditional or unconditional?

Or should this be part of collecting the initial information? i.e. if you specify 'collect-conditional-info' (whatever :) ) when collecting access information, it would be the same data structure,  just with an additional attribute for each access? Would you also need the info about statement (3) (i.e. the read and write there), in addition to having this info as part of the 'if' statement?

**Comment by sergisiso** [comment]
The place where I need if is fact one step above this, I am doing:
```
dep_tools = DependencyTools()
node.independent_iterations(dep_tools, ...)
for message in dep_tools.get_all_messages()
```
The current problem is that I cannot react differently because dependency messages are only the write-write one. When in reality I would like to do something different depending if they are exhaustive or not.

To a certain extend I don't mind how it is implemented, so feel free to use the simpler solution. I do not need the specific statement, as I am just looking for its existence and the symbol name.

**Comment by hiker** [comment]
@sergisiso , some additional questions where I am not sure about the best way to handle them:

1. Besides if-statements, what else should I make conditional - loops I would assume?

2. How should I handle multiple accesses? E.g.:
    ```
     if (a.ge.1) then
        c = 1
        if (a.eq.2) then
           c = 2
        endif
    endif
    ```
   You will only get accesses for the (outer) if statement reported. That's a read for `a`, but what about c? I could either add one unconditional read access, or add two accesses (first one unconditional, the second conditional)?

3. If you should decide to prefer having only a single access recorded for each variable, what about arrays used with different indices?
    ```
    if (a.ge.1) then
        c = b(i,j)
        if (a.eq.2) then
            c = b(i+1,j)
       endif
    endif
    ```
    We have one unconditional read access to `b(i,j)`, and one conditional read access to `b(i+1,j)`

**Comment by sergisiso** [comment]
> Besides if-statements, what else should I make conditional - loops I would assume?

Although it is technically true, I would probably not do loops for now as all the code I am processing has nested inner loops but in any case they are not run at least once. So it will be a lot of false positive for nothing. Maybe we can wait until we find a use case for it?

> How should I handle multiple accesses?
> If you should decide to prefer having only a single access recorded for each variable, what about arrays used with different indices?

I don't understand how this affects the multiple access and the indices. Wouldn't it be still the same implementation as we have now, but just adding a "_conditional" flag in the AccessInfo or different enum values in AccessType, whatever is easier. And then just be careful to populate them accordingly in the reference_accesses when calling `var_accesses.add_access(sig, ..., self, all_indices)`?

**Comment by sergisiso** [comment]
> I guess, the potential problem is if the user wants to use the collected information for further analysis.

I am not sure I follow all your reasoning but I think whether is conditional or not will always depend on the location that we are starting the query/analysis. For instance outer ifs are a common pattern in NEMO:
```fortran
if (feature_flag == .true.) then 
   do i=1,10
       field(i) = field(i) + scalar
   end
end if
```
And for analysing if this loop is parallel we don't consider 'field' a conditional access. But I think this is ok since we create a new VariablesAccessInfo every time we analyse a loop that starts from that node. IfBlock.reference_access could make this decisions (set it as conditional if its only in one branch) when its being found in a reference_access recursion. But I am aware I am just thinking about my use case here and I am unsure if there are other scenarios where we need this.

**Comment by hiker** [comment]
Just one additional question to confirm I am going in the right direction. In the following code:
```
if ()
   b(i+1) = 0
   b(i) = 0
else
  b(i) = 2
end if
```
then `b(i+1)` becomes a conditional write, the two `b(i)` writes are unconditional writes. But when evaluating all accesses to `b` (and not the individual accesses), `SingleVariableAccessInfo` for `b` will return that it is conditionally written?

**Comment by sergisiso** [comment]
Wouldn't "unconditional" take precedence here, if there is one conditional and one unconditional access to the same variables?

**Comment by hiker** [comment]
Oops - typo, yes, that's what I meant. As soon as any access is unconditional, all of them will be (when accessing the information about a variable) - sorry.

**Comment by hiker** [comment]
Arg, I am confused (and deleted my comment). I admit, I am not certain. What if the `i` and `i+1` are switched? We would have an unconditional write to `b(i+1)`, and conditional to `b(i)`? Would that change your opinion (because now the 'normal', 'non-stencil' access is conditional?

My initial thought is that in case of an array, the usage should indicate that there is at least one location that is conditionally written.

But I guess we can postpone this decision by providing better names 'higher level' function names instead of just `is_(un)conditional`: `at_least_one_conditional_write()` and `at_least_one_unconditional_write()` for `SingleVariableAccess`.

**Comment by hiker** [comment]
Hi @sergisiso ,

I've pushed my branch `2727_conditional_accesses` to get some quick feedback. TBH, I feel like the way we decided to go is not ideal (esp. with the impossible to handle situation of `if() then call somekernel(b) else b(i)=1; endif` - if somekernel results in a `readwrite` access, then we have a conditional read, and an unconditional write, and I can't set the first access (in the kernel) to both).

Modify the test `tets_variables_access_info_array_conditional` in tests/core/variables_access_info_test to do some tests, and look at the debug output that are in the branch 

Anything with `conditional 1` indicates a signature that is only on one of the two bodies (if or else). The other output (either just `conditional VARNAME` or `unconditional VARNAME` means the signature is in both branches, so it analyses the indices to get more details and tries to mark the conditional accesses correspondingly. This output is printed for each accesses, e.g. the line `unconditional write_if_else(i + 1)` is repeated three times, since there are three accesses to this variable with this index, and all are set to be unconditional.

Try to add your example code to this test, and see if the output makes sense.

My feeling is that this is just not good enough. I still hate that in the example:
```
          if (cond_var .ne. 1) then
              if (cond_var .eq. 2) then
                  write_if_else(i+1) = read_if_else(i)
              endif
              write_if_else(i+1) = read_if_else(1+i-1)
          else
              write_if_else(i) = read_if_else(i)
              write_if_else(i+1) = read_if_else(i)
          endif
```
the variable `write_if_else(i+1)` will be marked as unconditional in all three instances (even in the inner if-statement).

It still needs a lot of testing, documentation, and improvements I'd guess, but before spending more time, it would be great if you could give me some feedback if this is ok or not.

```
joerg@Joerg-Surface:~/work/psyclone/src/psyclone/tests/core$ pytest -s -k test_variables_access_info_array_conditional  ./variables_access_info_test.py 
/home/joerg/.local/lib/python3.10/site-packages/pep8.py:110: FutureWarning: Possible nested set at position 1
  EXTRANEOUS_WHITESPACE_REGEX = re.compile(r'[[({] | []}),;:]')
============================================================================== test session starts ===============================================================================
platform linux -- Python 3.10.12, pytest-7.4.4, pluggy-1.0.0
rootdir: /home/joerg/work/psyclone
configfile: setup.cfg
plugins: pylint-0.21.0, pep8-1.0.6, flakes-4.0.5, xdist-3.2.1, anyio-3.6.2, cov-4.0.0
collected 23 items / 22 deselected / 1 selected                                                                                                                                  

variables_access_info_test.py conditional 1 read_if(i)
conditional 1 write_if_else(i + 1)
conditional 1 i
conditional 1 i
conditional 1 i
conditional 1 i
conditional 1 write_if(i)
conditional 1 read_if_else(i)
===============================
<<<<<<<<<<<<<<<<<<<<<<<<< the above is for the inner if statement, which is indeed accessing i conditional <<<<<<<<<


conditional read_if(i)
conditional read_if(i)
-----------------------------
conditional 1 read_else(i)
conditional 1 cond_var
conditional 1 write_else(i)
===============================
unconditional write_if_else(i + 1)
unconditional write_if_else(i + 1)
unconditional write_if_else(i + 1)
-----------------------------
===============================
unconditional i
unconditional i
unconditional i
unconditional i
unconditional i
unconditional i
unconditional i
unconditional i
unconditional i
unconditional i
unconditional i
unconditional i
unconditional i
unconditional i
-----------------------------
===============================
conditional write_if(i)
conditional write_if(i)
-----------------------------
===============================
unconditional read_if_else(i)
unconditional read_if_else(i)
unconditional read_if_else(i)
unconditional read_if_else(i)
-----------------------------
%cond_var: READ, i: READ, %read_else: READ, %read_if: READ, read_if_else: READ, %write_else: WRITE, %write_if: WRITE, %write_if_else: WRITE
.
```

**Comment by sergisiso** [comment]
I used the `tets_variables_access_info_array_conditional` to test my code above where `my_val` is conditionally written:
```
  do i = 1, 10
      if (array(i) > 3) then
        my_val(1) = 1
        array(i) = my_val(1)
      else
        array(i) = my_val(1)
      endif
  end do
```

but I get:
```
===============================
unconditional my_val(1)
unconditional my_val(1)
unconditional my_val(1)
-----------------------------
===============================
unconditional i
unconditional i
-----------------------------
===============================
unconditional array(i)
unconditional array(i)
-----------------------------
array: READ+WRITE, i: READ, my_val: READ+WRITE
```
Is the problem that it is a conditional write + an unconditional read and this gets merged to unconditional read+write?

In the comment above regarding:
```
if ()
   b(i+1) = 0
   b(i) = 0
else
  b(i) = 2
end if
```
I said `Wouldn't "unconditional" take precedence here` but I was wrong, if they are 2 indices of the same symbol/signature in the same body the "conditional" should take precedence (as part of the access indices are conditional). Would that solve it?

Also thinking a bit more about the use-cases, the safe option is the "conditional" (not parallelise due to a value going to the next iteration, add the first in firstpirvate, ... ) so, in case of doubt I would tent to put "conditional" and just "unconditional" when we can guarantee it.

**Comment by hiker** [comment]
I had a look at your example, and there was indeed a typo. That seems to be better now. I've added your example (for now), and also renamed some variables to make that part of the code easier to understand.

For the rest of your comments I need a bit more time.

**Comment by sergisiso** [comment]
Note that I also added 2 xfail test in src/psyclone/tests/psyir/tools/dependency_tools_test.py with a similar snippet  and referencing this issue when was doing the array privatisation PR.
